src/data/repositories/ProgramLoggerD2Repository.ts

```diff
@@ -25,20 +27,45 @@ export class ProgramLoggerD2Repository implements LoggerRepository {
         this.messageId = dataElements.messageId;
         this.messageTypeId = dataElements.messageTypeId;
         this.organisationUnitId = organisationUnitId;
+        this.programStage = undefined;
+    }
+
+    async init(): Promise<void> {
+        this.programStage = await this.getProgramStage().toPromise();
     }
 
     log(log: DefaultLog): FutureData<void> {
-        return this.getProgramStage().flatMap(programStage => {
-            const d2EventProgram = this.mapLogToD2EventProgam({
+        if (!this.programStage) {
```

Comment:
And once refactored, all this this.programStage checks go away (the type will have no Maybe then)